lib/rule-tester/rule-tester.js

```diff
@@ -407,12 +408,75 @@ function assertErrorsProperty(errors, ruleName) {
 		}
 	}
 
+	const { requireMessage = false, requireLocation = false } =
+		assertionOptions;
+
 	if (isArray) {
 		assert.ok(
 			errors.length !== 0,
 			"Invalid cases must have at least one error",
 		);
+
+		for (const [number, error] of errors.entries()) {
+			if (typeof error === "string" || error instanceof RegExp) {
+				// Just an error message.
+				assert.ok(
+					requireMessage !== "messageId" && !requireLocation,
+					`errors[${number}] should be an object when 'assertionOptions.requireMessage' is 'messageId' or 'assertionOptions.requireLocation' is true.`,
+				);
+			} else if (typeof error === "object" && error !== null) {
+				/*
+				 * Error object.
+				 * This may have a message, messageId, data, line, and/or column.
+				 */
+
+				for (const propertyName of Object.keys(error)) {
+					assert.ok(
+						errorObjectParameters.has(propertyName),
+						`Invalid error property name '${propertyName}'. Expected one of ${friendlyErrorObjectParameterList}.`,
+					);
+				}
+
+				if (requireMessage === "message") {
+					assert.ok(
+						!hasOwnProperty(error, "messageId") &&
+							hasOwnProperty(error, "message"),
+						`errors[${number}] should specify 'message' (and not 'messageId') when 'assertionOptions.requireMessage' is 'message'.`,
+					);
+				} else if (requireMessage === "messageId") {
```

Comment:
ideally if the value is `messageId` we should ensure that all the placeholder values are also passed.

Reply:
In https://github.com/eslint/rfcs/pull/137 this was discussed, but it was excluded from the first iteration.
Search for `requireData`.